# Differential Gene-Expression Analysis with Machine Learning Using Dexamethasone Treatment Data

**Introduction** 

Worked with [Nityam Rathi](https://www.linkedin.com/in/nityam-rathi-75ab38128/), [Matt Eckelmeyer](https://www.linkedin.com/in/matthew-eckelmeyer-829b8b158/) and Kevin Coleman at University of Pittsburgh under supervision of [Dr. Junshu Bao](https://www.stat.pitt.edu/people/junshu-bao), Department of Statistics, University of Pittsburgh.  

The purpose of this project was to do a Differential Gene Expression Analysis on a Dexamethasone Treatment data set and incorporate Machine Learning to predict which genes could be differentially expressed. 

The datasets used are found under the National Center for Biotechnology Information (NCBI) Gene Expression Omnibus (GEO) accession number [GSE6711](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE6711). We specifically used files `GSM154262.txt, GSM154263.txt, GSM154264.txt, GSM154277.txt, GSM154278.txt, GSM154279.txt`. 

Dexamethasone specifically affects cell growth and can cause apoptosis (cell death) as specified [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5620202/). 

**Files** 

ðŸ“œ `Source Code.Rmd`

Contains the R Markdown code needed to create the final PDF file. 

ðŸ“œ `Dexamethasone Treatment Differential Gene Expression Analysis with Machine Learning.pdf`

Final PDF file outputted from knitting the source code. Contains all goals, steps, explanations of procedure for this research. 

**Contact Information** 

![interests](https://avatars1.githubusercontent.com/u/38919947?s=400&u=49ab1365a14fac78a91e425efd583f7a2bcb3e25&v=4)

Yogindra Raghav (YogiOnBioinformatics) 

yraghav97@gmail.com


